### DISCUSSION TITLE
Proposal: OPSS (Open Publication Synchronization Service)

### DISCUSSION THREAD

**Discussion by Fmstrat** [opds-community/drafts#49] [Ideas]
Hi everyone,

I'd like to make a proposal that may be of interest to you based on your backgrounds:

@janeczku - Author of Calibre-web
@HadrienGardeur - Maintainer of this repo and part of the OPDS community
@opds-community - The OPDS community
@geometer - Author or FBReader

Hopefully I've defined you all appropriately, but if not, my apologies. I'm sure there are others that make sense to pull into this conversion, and hopefully those already mentioned can reference them here.

While the OPDS spec is widely used (http://opds-spec.org/), one thing that has not yet been standardized in open readers is synchronization of read status, notes, bookmarks, and highlights. Many reader applications have this features (including FBReader), but while open, they are proprietary in-built systems.

My proposal is the creation of OPSS (Open Publication Synchronization Service) to accompany OPDS and hopefully be integrated into many of the solutions offering OPDS.

Initial considerations:
1. Web based, to ensure compatibility with existing OPDS solutions
2. JSON or XML response to allow for future addition of fields without breaking backwards compatibility
3. Back-end storage of data is up to the software developer, but guidance around DB structure or on-filesystem structures could be provided 
4. Guidance around authentication should be determined. (Shared user/pass with OPDS? API Key for public libraries? etc)
5. URL service location (always a subdirectory of OPDS to not require a separate URL?)
6. Should the spec allow for synchronization of items outside the OPDS library?

The following methods should be considered for implementation (naming is placeholder):
- **setposition**: Sets the current and furthest read position for 1 pub
- **getposition**: Gets the current and furthest read position 1 pub
- **getallpositions**: Gets all positions for multiple pubs (for quick synchronization of all books in the reader at load time for apps)
- **getmeta**: Gets both read positions, notes, bookmarks, and highlights 1 pub
- **setmeta**: Sets both read positions, notes, bookmarks, and highlights for 1 pub

Anyone interested in pursuing this?

----
**Add-ons**
Adding the below to the chain because I believe they are likely candidates for initial implementations (along with calibre-web from a server side) of a cross-platform solution (common platforms, actively developed, and FOSS).

@axet - Author of Book Reader
@babluboy - Author of Bookwork

----
- Relocated from: https://github.com/Feedbooks/opds-test-catalog/issues/1
- Similar to https://github.com/opds-community/drafts/issues/27
----
**Previously discussed with interest from @geometer (Author of FBReader)**
> (1) How to identify a book? Note that ePub is not the only book format.

Perhaps MD5 the file for a unique id?

> (2) How to record a position?

Count the words up the first word displayed on screen. That becomes the starting line for any reader, maybe? That also allows us to calculate percent read, and pages by defining average number of words per screen.

**Comment by HadrienGardeur** [comment]
> JSON or XML response to allow for future addition of fields without breaking backwards compatibility

I think such a spec should integrate with both OPDS 1.2 or 2.0 through a link, but I don't think that the "Synchronization Spec" would need both XML and JSON.

In 2019, I would simply default to JSON or JSON-LD.

> Back-end storage of data is up to the software developer, but guidance around DB structure or on-filesystem structures could be provided

IMO that's all out of scope for such a document.

> Guidance around authentication should be determined. (Shared user/pass with OPDS? API Key for public libraries? etc)

See https://drafts.opds.io/authentication-for-opds-1.0

> URL service location (always a subdirectory of OPDS to not require a separate URL?)

I'm not sure what you're talking about with that one. If you mean a "well-known location", I don't think that's the way to go.

The synchronization service should be discoverable using a link.

> Anyone interested in pursuing this?

NYPL has developed something like that for SimplyE that integrates with OPDS, might be worth having @leonardr chiming in as well.

**Comment by axet** [comment]
I'm using offline p2p synchronisation in my book reader. If you create compatible format json device to device synchronisation format I'm interested. Keep in mind synching across devices can create conflicts if book read by two devices and synchronisation software (create -conflict suffix files) and book reader (show conflict dialogs) should deal with it. I'm not interested in centralised only solution, not everyone want to setup a server.

But I store a bit more information about books such as:

```json
{
  "created": 1539355003077,  // added to library
  "last": 1553022686211, // last viewed
  "title": "Fallout_ Equestria", // can be renamed by user
  "position": [
    95, // paragraph
    0, // element
    0 // char
  ],
  "fontsize_6e01be161064d066": 45, // per device id font size, optimal to have per every device
  "bookmarks": [
    {
      "last": 1553022479332, // last change time for syncronization
      "name": "", // user name
      "text": "A pony’s PipBuck generates an E.F.S. (Eyes-Forward Sparkle) that will indicate direction and help gauge whether the ponies or creatures around you are hostile.", // actual highlight text
      "color": 0,
      "start": [
        97,
        223,
        0
      ],
      "end": [
        97,
        269,
        8
      ]
    },
    {
      "last": 1553022485681,
      "name": "",
      "text": "And, perhaps most impressively, a PipBuck can magically aid you in a fight for brief periods of time through use of the S.A.T.S. (Stable-Tec Arcane Targeting Spell).",
      "color": 0,
      "start": [
        97,
        271,
        0
      ],
      "end": [
        97,
        323,
        7
      ]
    }
  ]
}
```

**Comment by Fmstrat** [comment]
@axet If you have a moment, I have a couple of questions about your implementation:

- Out of curiosity, why did you choose to use paragraph, element, and char for positional look ups?
- Is this position the top-left most character displayed on the current screen?
- How are you identifying a unique book (or does that not matter due to the p2p nature of the tool)?


@HadrienGardeur 
> 
> 
> > JSON or XML response to allow for future addition of fields without breaking backwards compatibility
> 
> I think such a spec should integrate with both OPDS 1.2 or 2.0 through a link, but I don't think that the "Synchronization Spec" would need both XML and JSON.
> 
> In 2019, I would simply default to JSON or JSON-LD.
> 

Agreed, now that I've reviewed the 2.0 draft spec, JSON is definitely the way to go.

> > Back-end storage of data is up to the software developer, but guidance around DB structure or on-filesystem structures could be provided
> 
> IMO that's all out of scope for such a document.
> 

Agreed.

> > Guidance around authentication should be determined. (Shared user/pass with OPDS? API Key for public libraries? etc)
> 
> See https://drafts.opds.io/authentication-for-opds-1.0
> 

Makes sense to me.

> > URL service location (always a subdirectory of OPDS to not require a separate URL?)
> 
> I'm not sure what you're talking about with that one. If you mean a "well-known location", I don't think that's the way to go.
> 
> The synchronization service should be discoverable using a link.
> 

I guess the foundation of the question is, should OPSS be a separate service with a separate discoverable URL, or integrated into OPDS in a standard location.

> > Anyone interested in pursuing this?
> 
> NYPL has developed something like that for SimplyE that integrates with OPDS, might be worth having @leonardr chiming in as well.

Would certainly love as much input from interested parties as possible.

**Comment by HadrienGardeur** [comment]
It's also important to point out the work we're doing at Readium on locators, which could be the building block for such a service: https://readium.org/architecture/locators/

**Comment by leonardr** [comment]
We use the [Web Annotation Protocol](https://www.w3.org/TR/annotation-protocol/) to sync bookmarks and last reading position across devices. At a glance it covers all the use cases here, and it's a well-defined protocol with multiple independent implementations.

In particular, WAP defines a relation for discovery. Here's how we link to the annotation endpoint for a specific book in an OPDS 1.2 feed. Note the distinctive link relation and media type:

```
    <link href="https://circulation.librarysimplified.org/NYNYPL/annotations/Gutenberg%20ID/40278/" type="application/ld+json; profile=&quot;http://www.w3.org/ns/anno.jsonld&quot;" rel="http://www.w3.org/ns/oa#annotationService"/>
```

Since annotation documents are JSON-LD, I believe you can store whatever extra information you want to keep, bibliographic or otherwise.

The only extension we've had to make is a custom "motivation" (i.e. a reason why a person would create an annotation). We gave it the URI `http://librarysimplified.org/terms/annotation/idling` and it's used to indicate the last reading position -- a part of the book that is only special because it's the last thing that was on your screen when you closed the reader.

We don't promise to accept requests from any abitrary WAP client, but you can use WAP to extract all of your annotations from one of our servers.

As Hadrien points out (and I think Fmstrat is also getting at), another crucial piece of the puzzle that's currently missing is a reliable way to identify a specific place in a book. We use EPUB's [Content Fragment Identifier](http://www.idpf.org/epub/linking/cfi/epub-cfi.html) but they're not very reliable and we're hoping for something better to come out of Readium's "locators" project.

**Comment by majora2007** [comment]
There is another spec that is used by software of reading software called OPDS-PS and has support in quite a few software that is similar to this. This allows for streaming pages one by one to the reader and thus allows for the server to keep track of progress. 

My app, [Kavita](https://www.kavitareader.com/), has implemented this. We also have support from [Komga](https://komga.org/), [Panels](https://panels.app/) (client), [Codex](https://github.com/ajslater/codex/), and [LANraragi](https://github.com/Difegue/LANraragi)

https://github.com/anansi-project/opds-pse/blob/master/v1.2.md

**Comment by HadrienGardeur** [comment]
Hello @majora2007,

OPDS-PSE is an interesting mix of things:

- it provides a way to stream comics/manga page by page (image by image)
- there's also a built-in support for requesting a resized version of each page (can be useful to get a lower resolution version on smaller devices for example)
- and it provides minimal info for restoring progression

It's balancing out all these features with a single templated URI and a few attributes thanks to smart decisions and the linear nature of comics/manga.
For most of us working with offline support and packaged publications, we can't rely on the last element fetched on a remote server unfortunately.

There are restrictions and you could go beyond each of these features with various other standards out there, for instance:

- streaming page by page through a single URI template doesn't allow you to send information about spreads, which are handled natively supported in the [Divina profile](https://readium.org/webpub-manifest/profiles/divina) of the Readium Web Publication Manifest
- [IIIF Image API](https://iiif.io/api/image/3.0/) is better at allowing on-the-fly changes, such as resizing and cropping and transcoding to other image formats
- and the progression system could be broken by implementations that would preload part of the publications or fetch every image in lower resolution for a thumbnail view

It's a bit unfortunate that OPDS-PSE is based on OPDS 1.x since it could really benefit from the work done on OPDS 2.0 and RWPM:

- Atom links do not allow URI templates in OPDS 1.x, whereas we've added native support for them in OPDS 2.0
- RWPM and OPDS 2.0 have native support for contributor roles that are relevant for comics/manga: https://github.com/readium/webpub-manifest/blob/master/schema/metadata.schema.json#L74-L91
- RWPM and OPDS 2.0 allow publications to specify their reading progression direction, which is very useful for manga: https://github.com/readium/webpub-manifest/blob/master/schema/metadata.schema.json#L107
- you could easily mix and match the simplified URI template streaming approach with things that we're working on for Divina (support for Webtoons, panel by panel navigation or multi-lingual publications)

**Reply by gotson** [reply]
Hi @HadrienGardeur , i haven't had time to read the OPDS 2.0 draft yet, not any of the links you mentioned above.

Given we have very concrete use cases for the existing OPDS-PSE (which we took over from Ubooquity), how would you envision its successor in an OPDS 2.0 world?

You give many pointers and ideas above, but concretely, who would be designing/validating specs for that ? And would those specs be part of OPDS itself, or some kind of extensions?

I think our small Anansi OPDS-PSE community is quite interested in driving this towards a more sustainable future, but we don't have the context about OPDS 2.0 and where it would fit exactly.

**Reply by HadrienGardeur** [reply]
@gotson as hinted in discussions on the OPDS-PSE repo, I think that the best approach would be to compare with the work done in the [Readium community around the Divina profile](https://readium.org/webpub-manifest/profiles/divina) and see if we can merge our efforts.

Divina could cover multiple use cases:

- enhancing CBZ/CBR/EPUB files with better metadata for comics/manga
- and stream publications while allowing downloads as well

OPDS 2.0 inherits the RWPM model used for Divina as well, which means that we could hit two birds with one stone.

We hold calls in the Readium community every other week and could easily set up a channel on Readium's Slack to facilitate such discussions. cc @llemeurfr

**Reply by gotson** [reply]
Unfortunately the timing of the call doesn't work for my timezone, but I'll have a look at the opds 2.0 draft when I have some time on my hands!

**Reply by HadrienGardeur** [reply]
@gotson most of the people involved in Divina are based in Europe, we could easily schedule a one off call in the morning, which would probably work best for you. Would anyone else involved in OPDD-PSE like to join as well?

**Reply by gotson** [reply]
@HadrienGardeur a morning europe session would definitely work, i am in UTC+8.

i think @majora2007 is based in the USA, @ajslater probably also. @fillito is in Europe IIRC.

**Reply by fillito** [reply]
I'm based in Spain yep. But I work for a US company so my working hours are more or less US Eastern. I'm quite flexible though. With a few days' notice, I could do anything between 8:00 - 21:00 UTC

**Reply by HadrienGardeur** [reply]
I'll be in Madrid in early June @fillito, I'm heading there for a conference. Could be a good opportunity to meet and discuss this topic as well.

@gotson @fillito could you send me an email at hadrien@demarque.com? It will be easier to set up a call than Github.

**Reply by gotson** [reply]
@HadrienGardeur i finally read the OPDS 2.0 draft, and i was wondering whether the RWPM wouldn't already answer partly to page streaming, as it lists the different resources in a publication. Or did i misunderstood what that is and does?

**Reply by HadrienGardeur** [reply]
@gotson exactly, and we have a specialized RWPM profile for comics/manga specifically with that in mind: https://readium.org/webpub-manifest/profiles/divina

**Reply by HadrienGardeur** [reply]
@majora2007 @ajslater would you like to be on the call as well?

**Reply by majora2007** [reply]
Please add me @HadrienGardeur

**Reply by gotson** [reply]
> @gotson exactly, and we have a specialized RWPM profile for comics/manga specifically with that in mind: https://readium.org/webpub-manifest/profiles/divina

Thanks, I start to have a better idea of what you meant above in the initial post.

I don't think the resizing on the fly was ever implemented by clients. Opds-pse was started by Ubooquity, and the only client I think was Chunky on ios. Both are not developed anymore.

I didn't implement image resizing in Komga, I don't know if it is available in Kavita, nor whether Panels implements it.

I will most likely try to implement opds 2 in Komga, with support for the divina profile for publications.

@HadrienGardeur would you know whether Aldiko Next or Thorium supports the divina profile? That would definitely help me testing while developing.

**Reply by llemeurfr** [reply]
Thorium Reader support the Divina profile, yes. Note that Thorium does not read ebooks in streaming, but downloads them (and packages them if needed) before reading. Different use case.

**Reply by gotson** [reply]
> Thorium Reader support the Divina profile, yes. Note that Thorium does not read ebooks in streaming, but downloads them (and packages them if needed) before reading. Different use case.

It would be good enough to test and validate that my opds 2 feed in divina profile is valid 😊

**Reply by llemeurfr** [reply]
We'll be able to tidy up the parsing if there are some issues. Don't hesitate to exchange with us (can be via thorium-support at edrlab.org)

**Reply by ajslater** [reply]
Looking at the the Divina spec, the `readingOrder` section seems kind of heavyweight for most comic usage. It is interesting to be so explicit about the presentation of every page in a publication, and the panels schema seems useful, but I suspect for most comic book serving applications, a shorter, less expressive structure might be a simpler alternative.

e.g. off the top of my head:
```json
"links": [
    {"rel": "self", "href": "http://example.org/manifest.json", "type": "application/divina+json"},
    {"rel": "page", 
     "href": "http://example.org/{pageNum}/page.jpg", 
     "template": true, 
     "type": "image/jpeg", 
     "pageMin": 0, 
     "pageMax": 400 }
  ],
```

Does not require 400 readingOrder elements and the method of display could be handled by the [metadata](https://readium.org/webpub-manifest/schema/metadata.schema.json) and [presentation](https://readium.org/webpub-manifest/modules/presentation.html) schemas.

**Reply by gotson** [reply]
That would only work if all the pages have the same media type though, no?

**Reply by HadrienGardeur** [reply]
@ajslater @majora2007 could you send me an email at [hadrien@demarque.com](mailto:hadrien@demarque.com)? It will be easier to set up a call than Github.

**Comment by ajslater** [comment]
Yes.  But I feel that is by far the most common situation.

**Reply by majora2007** [reply]
I've seen regularly, esp in Webtoons, where there are a mix of content types. Usually a gif and pngs, but sometimes jpgs written in for cover images.

**Comment by ajslater** [comment]
Perhaps it would be nice to have the `readingOrder` specitivity for only the relatively rare pages that differ from the norm. Like covers are often to be displayed centrally, and within a comic book there are occasionally one or two pages that require a spread type view. But the vast majority of pages are displayed the same.

**Reply by gotson** [reply]
Just to understand, this simplification is to reduce the complexity on server side? I don't know how the clients would feel about this, maybe having the full details make it easier and reduce work on the client. Idk.

**Comment by ajslater** [comment]
There's a tiny one time penalty on the server side to do the page analysis
and then storage, but like, why do you want to send redundant information
over the wire every time you sever a publication? 25 to hundreds of json
fragments that all say "serve the next page exactly like the last one".
gzip or brotli isn't going to compress that.

**Comment by HadrienGardeur** [comment]
Let's move this discussion to the Divina/RWPM repo instead since we're no longer discussing OPDS or synchronization.
